### PR TITLE
[Merged by Bors] - feat: `finset` distributivity lemmas

### PR DESCRIPTION
**Pull request by YaelDillies**
Match https://github.com/leanprover-community/mathlib/pull/18611

[`data.finset.lattice`@`9d684a893c52e1d6692a504a118bfccbae04feeb`..`2d44d6823a96f9c79b7d1ab185918377be663424`](https://leanprover-community.github.io/mathlib-port-status/file/data/finset/lattice?range=9d684a893c52e1d6692a504a118bfccbae04feeb..2d44d6823a96f9c79b7d1ab185918377be663424)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Finset/Lattice.lean [hunk 1]
```diff
@@ -472,7 +473,7 @@ theorem inf_mem (s : Set α) (w₁ : ⊤ ∈ s) (w₂ : ∀ (x) (_ : x ∈ s) (y
 #align finset.inf_mem Finset.inf_mem
 
 @[simp]
-theorem inf_eq_top_iff (f : β → α) (S : Finset β) : S.inf f = ⊤ ↔ ∀ s ∈ S, f s = ⊤ :=
+protected theorem inf_eq_top_iff (f : β → α) (S : Finset β) : S.inf f = ⊤ ↔ ∀ s ∈ S, f s = ⊤ :=
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Lattice.lean:476]: This change is missing from `sup_eq_bot_iff `

File: Mathlib/Data/Finset/Lattice.lean [hunk 2]
```diff
@@ -547,8 +553,62 @@ theorem inf_sup_distrib_right (s : Finset ι) (f : ι → α) (a : α) :
...
+    Codisjoint a (s.inf f) ↔ ∀ ⦃i⦄, i ∈ s → Codisjoint a (f i) :=
+  @Finset.disjoint_sup_right αᵒᵈ _ _ _ _ _ _
+#align finset.codisjoint_inf_right Finset.codisjoint_inf_right
+
+protected theorem codisjoint_inf_left :
+    Codisjoint (s.inf f) a ↔ ∀ ⦃i⦄, i ∈ s → Codisjoint (f i) a :=
+  @Finset.disjoint_sup_left αᵒᵈ _ _ _ _ _ _
+#align finset.codisjoint_inf_left Finset.codisjoint_inf_left
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Lattice.lean:566]: This can be deleted.

File: Mathlib/Data/Finset/Lattice.lean [hunk 3]
```diff
@@ -547,8 +553,62 @@ theorem inf_sup_distrib_right (s : Finset ι) (f : ι → α) (a : α) :
...
+  @sup_inf_sup αᵒᵈ _ _ _ _ _ _ _ _
+#align finset.inf_sup_inf Finset.inf_sup_inf
+
 end OrderTop
 
+section BoundedOrder
+
+variable [BoundedOrder α] [DecidableEq ι]
+
+--TODO: Extract out the obvious isomorphism `(insert i s).pi t ≃ t i ×ˢ s.pi t` from this proof
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Lattice.lean:578]: Is `×ˢ` correct here?

File: Mathlib/Data/Finset/Lattice.lean [hunk 4]
```diff
@@ -547,8 +553,62 @@ theorem inf_sup_distrib_right (s : Finset ι) (f : ι → α) (a : α) :
...
+  refine'
+    ⟨fun h g hg =>
+      h (g i <| mem_insert_self _ _) (fun j hj => g j <| mem_insert_of_mem hj)
+        (hg _ <| mem_insert_self _ _) fun j hj => hg _ <| mem_insert_of_mem hj,
+      fun h a g ha hg => _⟩
+  -- TODO: This `have` must be named to prevent it being shadowed by the internal `this` in `simpa`
+  have aux : ∀ j : { x // x ∈ s }, ↑j ≠ i := fun j : s => ne_of_mem_of_not_mem j.2 hi
+  -- porting note: `simpa` doesn't support placeholders in proof terms
+  have := h (fun j hj => if hji : j = i then cast (congr_arg κ hji.symm) a
+      else g _ <| mem_of_mem_insert_of_ne hj hji) (fun j hj => ?_)
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Lattice.lean:597]: So `?_` doesn't work in a `simpa using`?

**Review by eric-wieser — APPROVED** [review]
bors d+

Some minor comments

File: Mathlib/Data/Finset/Lattice.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Data/Finset/Lattice.lean:578]: Was. Will update

File: Mathlib/Data/Finset/Lattice.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Data/Finset/Lattice.lean:597]: Indeed not

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**